### PR TITLE
Add a new csv to import soil layers settings

### PR DESCRIPTION
**Pull request by MostafaGomaa93**
### Description
relates #237
relates #241
relates #244 
relates https://github.com/EcoExtreML/STEMMUS_SCOPE_Processing/issues/103

### Checklist
- [x] Add a reference to related issues.
- [x] @mentions of the person or team responsible for reviewing proposed changes.
- [x] This pull request has a descriptive title.
- [x] Code is written according to the [guidelines](http://cnl.sogang.ac.kr/cnlab/lectures/programming/matlab/Richard_Johnson-MatlabStyle2_book.pdf).
- [x] The checks by [MISS_HIT style checker and linter](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/main/CONTRIBUTING.md#follow-matlab-style-guidelines), below the pull request, are successful (green).
- [x] Documentation is available.
- [x] Add changes to the [changelog file](CHANGELOG.md) under section `Unreleased`.
- [x] Model runs successfully.
- [x] Ask a meinatainer to re-generate exe file if matlab codes are changed. About how to create an exe file, see [exe readme](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/main/run_model_on_snellius/exe/README.md).

### CODE REVIEW

**Comment by MostafaGomaa93** [comment]
I think the `[+groundwater\calculateSoilLayerThickness](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/add_csv_for_soil_layers_settings/src/%2Bgroundwater/calculateSoilLayerThickness.m)` function is more related to soil settings rather than groundwater. Do you think it is better to move those [lines](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/a56e4627ebc83a48ab319fa9b5b3644050c46be7/src/%2Bgroundwater/calculateSoilLayerThickness.m#L10C1-L16C78) to the `getModelSettings.m` after this [line](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/a56e4627ebc83a48ab319fa9b5b3644050c46be7/src/%2Bio/getModelSettings.m#L80)?. but, then, we will need to copy them also after [this line here](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/a56e4627ebc83a48ab319fa9b5b3644050c46be7/src/STEMMUS_SCOPE.m#L64C1-L64C30) when the new csv file is activated

**Comment by MostafaGomaa93** [comment]
@BSchilperoort 
Do I need to explicit the exact variables I needed from those two structures ([`ForcingData`](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/a56e4627ebc83a48ab319fa9b5b3644050c46be7/src/STEMMUS_SCOPE_exe.m#L35C25-L35C36) and [`gwfluxes`](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/a56e4627ebc83a48ab319fa9b5b3644050c46be7/src/STEMMUS_SCOPE_exe.m#L30)) in the code? I already did that in [my comment in the BMI table](https://github.com/EcoExtreML/STEMMUS_SCOPE/issues/209#issuecomment-2178950772)?

File: src/+conductivity/+hydraulicConductivity/calculateDTheta_LLh.m
Comment (MostafaGomaa93) [inline src/+conductivity/+hydraulicConductivity/calculateDTheta_LLh.m:1]: correct wrong function name

**Comment by MostafaGomaa93** [comment]
@SarahAlidoost 
where shoud I put the new csv file? I attached here [input_soilLayThick.csv](https://github.com/user-attachments/files/16398828/input_soilLayThick.csv)

File: src/+conductivity/+hydraulicConductivity/calculateDTheta_LLh.m
Comment (SarahAlidoost) [reply src/+conductivity/+hydraulicConductivity/calculateDTheta_LLh.m:1]: nice!

File: src/+conductivity/+hydraulicConductivity/calculateDTheta_LLh.m
Comment (SarahAlidoost) [inline src/+conductivity/+hydraulicConductivity/calculateDTheta_LLh.m:1]:
```suggestion
function dtheta_llh = calculateDTheta_LLh(dtheta_uuh, theta_m, theta_uu, theta_ll, gamma_hh, SoilVariables, VanGenuchten, ModelSettings)
```

File: src/+io/checkModelRun.m
Comment (SarahAlidoost) [inline src/+io/checkModelRun.m:1]: Please remove this function, see my comment below.

File: src/STEMMUS_SCOPE_exe.m
Comment (SarahAlidoost) [inline src/STEMMUS_SCOPE_exe.m:32]: why "Trap" is added here?

File: src/+io/readSoilLayerSettings.m
Comment (SarahAlidoost) [inline src/+io/readSoilLayerSettings.m:3]:
```suggestion
    SoilLayerSettings.NL = soildata(end, 1);
    SoilLayerSettings.ML = SoilLayerSettings.NL;
```

File: src/STEMMUS_SCOPE.m
Comment (SarahAlidoost) [inline src/STEMMUS_SCOPE.m:54]:
```suggestion
        SoilLayerSettings = io.readSoilLayerSettings(soilLayersFile);
```

File: src/STEMMUS_SCOPE.m
Comment (SarahAlidoost) [inline src/STEMMUS_SCOPE.m:65]:
```suggestion
        % Override the default settings
        fields = fieldnames(SoilLayerSettings);
        for i = 1:numel(fields)
            ModelSettings.(fields{i}) = SoilLayerSettings.(fields{i});
        end
    end

    % Calculate other settings
    ModelSettings.NN = ModelSettings.NL + 1;
    ModelSettings.mN = ModelSettings.NN + 1;
    ModelSettings.mL = ModelSettings.NL + 1;
    ModelSettings.nD = 2;
```

Comment (SarahAlidoost) [reply src/STEMMUS_SCOPE.m:65]: @MostafaGomaa93 can you please remove [these lines](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/18c39d3ac37a85579c2b5f67c78932ff11733755/src/%2Bio/getModelSettings.m#L77-L80) from `getModelSettings`

File: src/STEMMUS_SCOPE.m
Comment (SarahAlidoost) [inline src/STEMMUS_SCOPE.m:805]:
```suggestion
```

Comment (SarahAlidoost) [reply src/STEMMUS_SCOPE.m:805]: Please avoid adding `disp` messages to the main loop of the model. Doing so will increase the size of the command output sometimes to GB, especially when processing long time series for multiple locations on HPC, and can cause issues when running the model in parallel. If you need to capture this information, it should be logged to a separate file with a specific filename. or you can write a test script under directory `tests`.

**Comment by SarahAlidoost** [comment]
> @SarahAlidoost I think the [`+groundwater\calculateSoilLayerThickness.m`](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/add_csv_for_soil_layers_settings/src/%2Bgroundwater/calculateSoilLayerThickness.m) function is more related to soil settings than the groundwater calculations. Do you think it is better to move those [lines](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/a56e4627ebc83a48ab319fa9b5b3644050c46be7/src/%2Bgroundwater/calculateSoilLayerThickness.m#L10C1-L16C78) to the `getModelSettings.m` after this [line](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/a56e4627ebc83a48ab319fa9b5b3644050c46be7/src/%2Bio/getModelSettings.m#L80)?. but, then, we will need to copy them also after [this line here](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/a56e4627ebc83a48ab319fa9b5b3644050c46be7/src/STEMMUS_SCOPE.m#L64C1-L64C30) when the new csv file is activated

For readability, let's keep the calculations in its own function `calculateSoilLayerThickness`. This function can be placed either in the folder `+io` or in a new folder like `+soil.`

**Review by SarahAlidoost — CHANGES_REQUESTED** [review]
@MostafaGomaa93 thanks for fixing the ModelSettings. See my comments.

File: src/STEMMUS_SCOPE.m
Comment (MostafaGomaa93) [reply src/STEMMUS_SCOPE.m]: > @MostafaGomaa93 can you please remove [these lines](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/18c39d3ac37a85579c2b5f67c78932ff11733755/src/%2Bio/getModelSettings.m#L77-L80) from `getModelSettings`

Instead of removing them, we can keep them in the [`getModelSettings`](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/18c39d3ac37a85579c2b5f67c78932ff11733755/src/%2Bio/getModelSettings.m#L77-L80) and add them also to the [`ReadSoilLayerSettings`](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/fbb938b029191da63259829360bcf03d113d0adf/src/%2Bio/readSoilLayerSettings.m#L6C3-L9C30). Then, we don't need to add them in the [`main function`](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/77218e8e498ce26515b532a85a22c130cf7cd9b8/src/STEMMUS_SCOPE.m#L61C2-L65C26)

**Comment by MostafaGomaa93** [comment]
> For readability, let's keep the calculations in its own function `calculateSoilLayerThickness`. This function can be placed either in the folder `+io` or in a new folder like `+soil.`

Okay, I move it to the `+io`

**Comment by MostafaGomaa93** [comment]
> You can add it to a new folder named `example_data` in the root directory. Also, please add one line in [STEMMUS_SCOPE_on_CRIB.md](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/main/docs/STEMMUS_SCOPE_on_CRIB.md#dataflow-of-stemmus_scope-on-crib) and [STEMMUS_SCOPE_on_Snellius.md](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/main/docs/STEMMUS_SCOPE_on_Snellius.md#dataflow-of-stemmus_scope-on-snellius) under section **Dataflow** to explain where numbers in csv file come from and how users can use the csv file, for example, they have to copy this file to the model's input directory, ....

Okay, I add an explanation as you suggest. Please have a look and let me know if it is clear or something is still missing

File: src/STEMMUS_SCOPE.m
Comment (MostafaGomaa93) [reply src/STEMMUS_SCOPE.m:797]: > If you need to capture this information, it should be logged to a separate file with a specific filename. or you can write a test script under directory `tests`.

Could you please guide me how to log this information to a separate file with a specific filename? I indeed need this info, especially in long time-series models to have a track of the running time

File: src/STEMMUS_SCOPE_exe.m
Comment (SarahAlidoost) [reply src/STEMMUS_SCOPE_exe.m:32]: For adding variables to BMI, two files should be changed, the `STEMMUS_SCOPE_exe.m` in this repo and `varaible_refence.py` in pystemmusscope. This is beyond the scope of issue #241 and should be done in a separate pull request. You can also add a comment about it in issue https://github.com/EcoExtreML/STEMMUS_SCOPE/issues/209.

Comment (SarahAlidoost) [inline src/STEMMUS_SCOPE_exe.m:32]:
```suggestion

```

File: src/STEMMUS_SCOPE.m
Comment (SarahAlidoost) [reply src/STEMMUS_SCOPE.m]: This way you are duplicating code which is not recommended. Please avoid duplicating the code and fix it as suggested.

**Comment by SarahAlidoost** [comment]
@Crystal-szj this pull request refactors the codes of `ModelSettings`, related to your issue https://github.com/EcoExtreML/STEMMUS_SCOPE/issues/246. Now `ModelSettings = io.getModelSettings()` is **not** called in several scripts but instead, the input argument `ModelSettings` is passed as an input argument to the functions requiring model settings. This way, the `ModelSettings` can be changed from outside of the model. I have already reviewed this pull request. Can you please help to re-generate exe file on snellius using this branch? thanks.

**Review by SarahAlidoost — COMMENTED** [review]
@MostafaGomaa93 Thanks for the implementation :+1: . Once exe file is added, it can be merged.

**Comment by Crystal-szj** [comment]
> @Crystal-szj this pull request refactors the codes of `ModelSettings`, related to your issue #246. Now `ModelSettings = io.getModelSettings()` is **not** called in several scripts but instead, the input argument `ModelSettings` is passed as an input argument to the functions requiring model settings. This way, the `ModelSettings` can be changed from outside of the model. I have already reviewed this pull request. Can you please help to re-generate exe file on snellius using this branch? thanks.

Sure. I'll recreate it and let you know.

**Comment by Crystal-szj** [comment]
Hi @SarahAlidoost @MostafaGomaa93, I have re-generated the executable file.

**Comment by SarahAlidoost** [comment]
> Hi @SarahAlidoost @MostafaGomaa93, I have re-generated the executable file. The merging is still blocked. Should I merge this branch into the main branch directly?

@Crystal-szj Great, Thanks :+1:  the reason that merging is blocked is that the pull request is waiting for the approval of another reviewer (in this case me :smile: ). I will go ahead with the pull request.

**Comment by SarahAlidoost** [comment]
@MostafaGomaa93 and @Crystal-szj I noticed that the model run isnot successful. I fixed the code in this [commit](https://github.com/EcoExtreML/STEMMUS_SCOPE/pull/243/commits/c86ce7ad3c8823f46e23e91228e6edc4de068b3a) and re-generate exe file.

**Review by SarahAlidoost — APPROVED** [review]

**Comment by MostafaGomaa93** [comment]
> @MostafaGomaa93 and @Crystal-szj I noticed that the model run isnot successful. I fixed the code in this [commit](https://github.com/EcoExtreML/STEMMUS_SCOPE/pull/243/commits/c86ce7ad3c8823f46e23e91228e6edc4de068b3a) and re-generate exe file.

Thanks @SarahAlidoost for this leftover fix

**Pull request merged**
